### PR TITLE
Add RealmList.removeAllFromRealm and Realm.clear

### LINKED ISSUE #1560: Add `Realm.delete(RealmList)` and `Realm.delete(RealmResults) to remove RealmObjects
Cocoa binding has the similar functions
See https://realm.io/docs/swift/latest/api/Classes/Realm.html#/s:FC10RealmSwift5Realm6deleteuRq_Ss12SequenceTypedqqq_S1_9GeneratorSs13GeneratorType7ElementCS_6Object_FS0_Fq_T_

Currently, All objects in `RealmResults` can be deleted by `RealmResults.clear()`
But there is no function to delete all objects from `RealmList`

### PR DESCRIPTION
**Pull request by beeender**
* Add LinkView.removeAllTargetRows.
* Add Realm.removeAll.
* Add Realm.remove(RealmObject).
* Add Realm.remove(Iterable<RealmObeject>).
* Javadoc & test case update.

Close #1560

### CODE REVIEW

**Comment by beeender** [comment]
@realm/java

File: changelog.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@
  * Fixed an error occurring during test and connectedCheck of unit test example (#1934).
  * Added RealmQuery.isNotEmpty() (#2025). (Thank you @stk1m1)
  * Improved performance of RealmList#contains() (#897).
+ * Added Realm.remove(RealmObject), Realm.remove(Iterable) and Realm.removeAll() (#1560).
```
Comment (dalinaum) [inline changelog.txt]: How about rearranging items.

```
  * Added RealmQuery.isNotEmpty() (#2025). (Thank you @stk1m1)
  * Added Realm.remove(RealmObject), Realm.remove(Iterable) and Realm.removeAll() (#1560).
  * Improved performance of RealmList#contains() (#897).
```

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 2]
```diff
@@ -535,6 +534,45 @@ public RealmSchema getSchema() {
...
+    /**
+     * Remove the given objects from Realm which they are associated with.
+     */
+    static <E extends RealmObject> void remove(Iterable<E> realmObjects) {
+        if (realmObjects instanceof RealmResults)  {
+            ((RealmResults) realmObjects).clear();
+        } else if (realmObjects instanceof RealmList && ((RealmList) realmObjects).isValid()) {
+            ((RealmList) realmObjects).removeAllFromRealm();
+        } else {
+            for (RealmObject object : realmObjects) {
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: If `realmObjects` is a `RealmList` and `realmObjects.isValid()` return `false`, you'll get into this block, right?

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 3]
```diff
@@ -535,6 +534,45 @@ public RealmSchema getSchema() {
...
+    void removeAll() {
+        checkIfValid();
+        for (RealmObjectSchema objectSchema : schema.getAll()) {
+            schema.getTable(objectSchema.getClassName()).clear();
+        }
+    }
+
+    /**
+     * Remove the given object from Realm which it is associated with.
+     */
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: Can't `removeFromRealm()` throw anything?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 4]
```diff
@@ -1173,6 +1173,48 @@ public void clear(Class<? extends RealmObject> clazz) {
...
+    @Override
+    public void removeAll() {
+        super.removeAll();
+    }
+
+    /**
+     * Removes the object from the Realm it is currently associated to. This method is equivalent to
+     * {@link RealmObject#removeFromRealm()}.
+     * <p>
+     * After this method is called the object will be invalid and any operation (read or write) performed on it will
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Missing `.` at the end.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 5]
```diff
@@ -1173,6 +1173,48 @@ public void clear(Class<? extends RealmObject> clazz) {
...
+     *
+     * @throws IllegalStateException if the corresponding Realm is closed or in an incorrect thread.
+     * @param realmObject the {@link RealmObject} to be removed from its corresponding Realm.
+     */
+    public static void remove(RealmObject realmObject) {
+        BaseRealm.remove(realmObject);
+    }
+
+    /**
+     * Remove a collection of RealmObjects from the underlying Realm.
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: "Remove" -> "Removes"

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 6]
```diff
@@ -1173,6 +1173,48 @@ public void clear(Class<? extends RealmObject> clazz) {
...
+     * <p>
+     * If the realmObjects is an instance of {@link RealmResults}, this method is equivalent to
+     * {@link RealmResults#clear()}. RealmObjects in the list will be removed from the list and the underlying Realm.
+     * <p>
+     * If the realmObjects is an instance of {@link RealmList}, RealmObjects in the list will be removed from the list
+     * and the underlying Realm.
+     * <p>
+     * If the realmObjects is other type of collections, the RealmObjects in the collection will be removed from the
+     * underlying Realm, but the collection will stay unchanged. All the RealmObject in the collection becomes invalid.
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: "becomes" -> "become"

File: realm/realm-library/src/main/java/io/realm/DynamicRealm.java [hunk 7]
```diff
@@ -129,6 +129,14 @@ public void clear(String className) {
     }
 
     /**
+     * {@inheritDoc}
+     */
+    @Override
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/DynamicRealm.java]: no need to override if you just need to call super

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 8]
```diff
@@ -1173,6 +1173,48 @@ public void clear(Class<? extends RealmObject> clazz) {
         getTable(clazz).clear();
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: no need to override if you just need to call super

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 9]
```diff
@@ -1173,6 +1173,48 @@ public void clear(Class<? extends RealmObject> clazz) {
...
+    public void removeAll() {
+        super.removeAll();
+    }
+
+    /**
+     * Removes the object from the Realm it is currently associated to. This method is equivalent to
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: > associated to

_associated with_ sounds more accurate?

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 10]
```diff
@@ -300,6 +300,21 @@ public E remove(int location) {
     }
 
     /**
+     * Remove all elements from this list and delete them from the corresponding Realm.
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmList.java]: > Remove

Removes

**Comment by nhachicha** [comment]
what happens when a pending insert transaction is running in the background. I guess  `removeAll` doesn't guarantee everything will be erased ( that's the user responsibility to cancel async transac & make sure the call return before calling `removeAll`

**Comment by beeender** [comment]
@nhachicha Yep, i think so. `removeAll` still needs to be called in the transaction and it will be user's responsibility to ensure the sequence of inserting and clearing.

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 8 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: I thought i should use `@Override` whenever it is possible?

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: You should, but in this case where your override isn't doing anything. You can just leave it out, both editor autocomplete + javadoc will find it correctly.

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java [hunk 11]
```diff
@@ -69,6 +71,8 @@ private void populateTestRealm(DynamicRealm realm, int objects) {
             allTypes.setFloat(AllTypes.FIELD_FLOAT, 1.234567F + i);
             allTypes.setString(AllTypes.FIELD_STRING, "test data " + i);
             allTypes.setLong(AllTypes.FIELD_LONG, i);
+            allTypes.getList(AllTypes.FIELD_REALMLIST).add(realm.createObject("Dog"));
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java]: Why not use the constants `CLASS_CAT` and `CLASS_DOG`?

File: changelog.txt [hunk 12]
```diff
@@ -3,6 +3,7 @@
  * BREAKING CHANGE: DynamicRealm.executeTransaction() now throws IllegalArgumentException instead of silently accepting a null Transaction object.
  * Fixed an error occurring during test and connectedCheck of unit test example (#1934).
  * Added RealmQuery.isNotEmpty() (#2025). (Thank you @stk1m1)
+ * Added Realm.remove(RealmObject), Realm.remove(Iterable) and Realm.removeAll() (#1560).
```
Comment (cmelchior) [inline changelog.txt]: "Added Realm.clear() and RealmList.removeAllFromRealm() (#1560)."

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java [hunk 13]
```diff
@@ -882,4 +886,42 @@ public void run() {
         }
     }
 
+    public void testClearAll() {
+        realm.beginTransaction();
+        realm.createObject(CLASS_ALL_TYPES);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java]: I would move these constants to their class instead, like "Alltypes.CLASS_NAME", some of them already exists.

File: realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java [hunk 14]
```diff
@@ -979,4 +981,110 @@ public void testSettingListClearsOldItems() {
...
+        Dog dog2 = new Dog();
+
+        list.add(dog1);
+        list.add(dog2);
+
+        testRealm.beginTransaction();
+        try {
+            list.removeAllFromRealm();
+            fail("Cannot remove a list with a standalone object in it!");
+        } catch (IllegalStateException e) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: Should this be an IllegalArgumentException instead?

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java (hunk 13 above)
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java]: I will do that when clean up this test with `TestRealmConfigurationFactory`

Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTest.java]: :+1:

File: realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java (hunk 14 above)
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: No, the exception is sent from `RealmObject` which is in a illegal state that it is not managed by Realm.

Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: Ah, good point :+1:

**Comment by beeender** [comment]
retest this please

**Comment by cmelchior** [comment]
retest this please

**Comment by cmelchior** [comment]
Retest this please

**Comment by beeender** [comment]
Retest this please

**Comment by beeender** [comment]
retest this please

**Comment by cmelchior** [comment]
@beeender Just noticed we have a `Realm.clear(class)`. Should we rename `clear()` to `clearAll()` instead or is `clear()` and `clear(class)` good enough?

**Comment by beeender** [comment]
@cmelchior I don't have strong opinion on both names. We might be able to have another function which is close to the word `clear` is clear everything, objects and schemas. Maybe that fits `clearAll` better? Currently it is easy for user to figure out the `clear()` is just clear all objects because of `clear(String classname)` is there. Right? Any other good reason that we shall use `clearAll` here?

**Comment by beeender** [comment]
retest this please

**Comment by cmelchior** [comment]
`clear()` fits the List API better, while `clearAll()` is a perhaps a bit easier to read if you don't know the  method exists on the List interface. I think I'm fine with `clear()`, I just noticed it when doing something else.

**Comment by beeender** [comment]
retest this please

**Comment by beeender** [comment]
retest this please and god bless me

File: realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java [hunk 15]
```diff
@@ -979,4 +981,110 @@ public void testSettingListClearsOldItems() {
...
+        try {
+            dogs.removeAllFromRealm();
+            fail("removeAllFromRealm should be called in a transaction.");
+        } catch (IllegalStateException e) {
+            assertEquals("Changing Realm data can only be done from inside a transaction.", e.getMessage());
+        }
+    }
+
+    @Test
+    public void testRemoveAllFromListStandaloneObjectShouldThrow() {
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: What does this test?

* IllegalStateException will be thrown if the `RealmList` is not managed.
* IllegalStateException will be thrown if the `RealmList` contains standalone object.

I think we should split this into two tests.

Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: oh, `testRealmRemoveAllNotManagedList already` exists and it does not throw the `IllegalStateException`.

so, this tests **IllegalStateException will be thrown if the RealmList contains standalone object.**

Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/RealmListTest.java]: "IllegalStateException will be thrown if the RealmList is not managed." This doesn't throw if every object is managed.
See testRealmRemoveAllNotManagedList
"IllegalStateException will be thrown if the RealmList contains standalone object." This is what this test do.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 16]
```diff
@@ -535,6 +534,18 @@ public RealmSchema getSchema() {
     }
 
     /**
+     * Removes all objects from this Realm.
+     *
+     * @throws IllegalStateException if the corresponding Realm is closed or in an incorrect thread.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: in -> on

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 17]
```diff
@@ -263,11 +263,11 @@ public void move(int oldPos, int newPos) {
     }
 
     /**
-     * Removes all elements from this list, leaving it empty.
-     *
+     * Removes all elements from this list, leaving it empty. This method doesn't remove the objects from the Realm.
      * @throws IllegalStateException if Realm instance has been closed or parent object has been removed.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmList.java]: Missing whitespace between section and throws

**Comment by cmelchior** [comment]
Apart from a few Javadoc comments this looks good :+1:

**Comment by beeender** [comment]
retest this please and don't fail!

**Comment by dalinaum** [comment]
retest this please

**Comment by kneth** [comment]
:+1:

**Pull request merged**